File: C++/String/39_unoccupied_cumputers.cpp
/*
Idea:
iterating string & keeping track of each character, if it's already passed or not.
& keeping count of remaining computers, updating as character comes or leaves.
*/


#include<bits/stdc++.h>
using namespace std;



class Solution
{
public:
    int solve(int n, string seq)
    {
        // seen[i] = 0, indicates that customer 'i' is not in cafe
        // seen[1] = 1, indicates that customer 'i' is in cafe but computer is not assigned yet.
        // seen[2] = 2, indicates that customer 'i' is in cafe and  has occupied a computer.

        char seen[26] = { 0 }; // array to store entry/exit of chars
        int res = 0, occupied = 0;

        for (int i = 0; seq[i]; i++)
        {
            int ind = seq[i] - 'A';

            if (seen[ind] == 0) // if char hasn't entered
            {
                seen[ind] = 1;
                if (occupied < n) // if computers are available
                {
                    occupied++;
                    seen[ind] = 2;
                }
                else
                    res++;
            }
            else
            {
                if (seen[ind] == 2) // if char has left
                    occupied--;
                seen[ind] = 0;
            }
        }
        return res;
    }
};


int main() {
    int t;
    scanf("%d", &t);
    while (t--)
    {

        int n;
        scanf("%d", &n);


        cin.ignore();
        string s = ;
        getline(cin, s);

        Solution obj;
        int res = obj.solve(n, s);

        cout << res << endl;

    }
}

File: Python/String/39_unoccupied_cumputers.py
"""
Idea:
iterating string & keeping track of each character, if it's already passed or not.
& keeping count of remaining computers, updating as character comes or leaves.
"""


class Solution:
    def solve(self, n, seq):
        # seen[i] = 0, indicates that customer 'i' is not in cafe
        # seen[1] = 1, indicates that customer 'i' is in cafe but computer is not assigned yet.
        # seen[2] = 2, indicates that customer 'i' is in cafe and has occupied a computer.

        seen = [0] * 26  # List to store entry/exit of characters
        res = 0
        occupied = 0

        for i in range(len(seq)):
            ind = ord(seq[i]) - ord('A')

            if seen[ind] == 0:  # If character hasn't entered
                seen[ind] = 1
                if occupied < n:  # If computers are available
                    occupied += 1
                    seen[ind] = 2
                else:
                    res += 1
            else:
                if seen[ind] == 2:  # If character has left
                    occupied -= 1
                seen[ind] = 0

        return res


for _ in range(int(input())):
    n = int(input())
    s = input().strip()
    ob = Solution()
    ans = ob.solve(n, s)
    print(ans)
